src/EventSubscriber/OrderEventSubscriber.php

```diff
@@ -0,0 +1,198 @@
+<?php
+
+namespace Drupal\commerce_stock\EventSubscriber;
+
+use Drupal\commerce_order\Event\OrderEvents;
+use Drupal\commerce_order\Event\OrderEvent;
+use Drupal\commerce_order\Event\OrderItemEvent;
+use Drupal\commerce_stock\StockServiceManagerInterface;
+use Drupal\state_machine\Event\WorkflowTransitionEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Performs stock transactions on order and order item events.
+ */
+class OrderEventSubscriber implements EventSubscriberInterface {
+
+  /**
+   * The stock service manager.
+   *
+   * @var \Drupal\commerce_stock\StockServiceManagerInterface
+   */
+  protected $stockServiceManager;
+
+  /**
+   * Constructs a new OrderReceiptSubscriber object.
+   *
+   * @param \Drupal\commerce_stock\StockServiceManagerInterface $stock_service_manager
+   *   The stock service manager.
+   */
+  public function __construct(StockServiceManagerInterface $stock_service_manager) {
+    $this->stockServiceManager = $stock_service_manager;
+  }
+
+  /**
```

Comment:
This is brilliant and I can not think of any normal case it does not cover and having it event driven results in the ability to do reliable back-end ordering. Only issues I am aware of are:
1) Some sites may not want stock automatically returned on cancel/deleted.. order as this may follow a different process.
2) for other it may not be enough as they may want to reserve stock on add to cart.
Not realty worried about (2) as this is more then we had 
We can easily add config option for (1) at a later point.